packages/contracts-bedrock/src/L2/OptimismSuperchainERC20.sol

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.25;
+
+import { IOptimismSuperchainERC20 } from "src/L2/IOptimismSuperchainERC20.sol";
+import { ERC20 } from "@solady/tokens/ERC20.sol";
+import { IL2ToL2CrossDomainMessenger } from "src/L2/IL2ToL2CrossDomainMessenger.sol";
+import { ISemver } from "src/universal/ISemver.sol";
+import { Predeploys } from "src/libraries/Predeploys.sol";
+
+/// @notice Thrown when attempting to relay a message and the function caller (msg.sender) is not
+/// L2ToL2CrossDomainMessenger.
+error CallerNotL2ToL2CrossDomainMessenger();
+
+/// @notice Thrown when attempting to relay a message and the cross domain message sender is not this
+/// OptimismSuperchainERC20.
+error InvalidCrossDomainSender();
+
+/// @notice Thrown when attempting to mint or burn tokens and the function caller is not the StandardBridge.
+error OnlyBridge();
+
+/// @notice Thrown when attempting to mint or burn tokens and the account is the zero address.
+error ZeroAddress();
+
+/// @custom:proxied
+/// @title OptimismSuperchainERC20
+/// @notice OptimismSuperchainERC20 is a standard extension of the base ERC20 token contract that unifies ERC20 token
+///         bridging to make it fungible across the Superchain. This construction allows the L2StandardBridge to burn
+///         and mint tokens. This makes it possible to convert a valid OptimismMintableERC20 token to a SuperchainERC20
+///         token, turning it fungible and interoperable across the superchain. Likewise, it also enables the inverse
+///         conversion path.
+///         Moreover, it builds on top of the L2ToL2CrossDomainMessenger for both replay protection and domain binding.
+contract OptimismSuperchainERC20 is IOptimismSuperchainERC20, ERC20, ISemver {
+    /// @notice Address of the L2ToL2CrossDomainMessenger Predeploy.
+    address internal constant MESSENGER = Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER;
+
+    /// @notice Address of the StandardBridge Predeploy.
+    address internal constant BRIDGE = Predeploys.L2_STANDARD_BRIDGE;
+
+    /// @notice Address of the corresponding version of this token on the remote chain.
```

Comment:
This is good altho we will likely need to change some things so that the implementation can live behind a 1967 beacon proxy, ie no immutables. We will also need to determine the specifics of how the proxies are deployed, ie are they clones with immutable args style?

Reply:
@tynes Yes, you are right. We will do a bit of research and adapt accordingly.